dotnet/src/InternalUtilities/src/Diagnostics/ModelDiagnostics.cs

```diff
@@ -109,16 +119,23 @@ public static bool IsModelDiagnosticsEnabled()
     }
 
     #region Private
-    private static void AddOptionalTags(Activity? activity, PromptExecutionSettings? executionSettings)
+    private static void AddOptionalTags<T>(Activity? activity, T? executionSettings) where T : PromptExecutionSettings
     {
-        if (activity is null || executionSettings?.ExtensionData is null)
+        if (activity is null || executionSettings is null)
+        {
+            return;
+        }
+
+        // Serialize and deserialize the execution settings to get the extension data
+        var deserializedSettings = JsonSerializer.Deserialize<PromptExecutionSettings>(JsonSerializer.Serialize(executionSettings));
```

Comment:
Is this always going to do the right thing? i.e. the exact same names are used for all execution settings for MaxToken, Temperature, and TopP?

Reply:
Yes, so far. We just need to be very careful about this. I am planning to add unit tests for all the model diagnostics code.